keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,640 @@
+
+## Motivation
+
+In practice, workloads may scale vertically up to a certain point but then must partition horizontally. The system scales up and down horizontally with partition and at the same time provides a software layer to manage key attributes iops/throughput independently from capacity.
+
+Currently after CreateVolume with provider specific parameters pass in storage class, Kubernetes does not handle update of QoS related parameters. The only way to achieve this is to use the cloud provider's API directly to modify iops and throughput. This creates frictions of managing storage resources on k8s. Allowing Resources to be changed without managing it through different provider's APIs directly simplifies this whole process. The principle here is that IO provisioning and volume conversions are treated as two separate problems.
+
+### Goals
+
+* Cloud-provider and storage vendor independent specification of basic QoS parameters
+* QoS enforcement performed by the storage
+* Specifying QoS at volume creation
+* Changing QoS parameters on-line for existing volumes
+* Specify quotas in a similar way to that of volume capacity
+
+### Non-Goals
+
+* OS-level IO QoS, for example cgroup policy
+* Inter-pod QoS. This KEP is restricted to what QoS can be defined on a volume.
+* Volume QoS classes. This proposal is for fine-grained provisioned IO control, and not for defining general classes of volume performance. Currently, storage vendors may accomplish this by adding custom parameters to a storage class. A future KEP may codify such behavior with common storage class parameter names, but that is out of scope of this KEP.
+* Scheduling based on total IO limits of a node. For example, total throughput for many cloud provider network attached storage volumes is limited by instance network bandwidth and may prevent achieving the provisioned IO. Scheduling to mitigate these factors is out of scope but it should be considered as a future extension of this KEP.
+
+## Proposal
+
+### Kubernetes API
+Add two well-known ResourceName keys to the existing ResourceList type used in PersistentVolumeClaim.Spec.Resources.{Limits,Requests} and PersistentVolume.Spec.Capacity:
+* (ReadWrite) IOPS independent from capacity
+* (ReadWrite) Throughput
+These values can be changed on a PersistentVolumeClaim, and will be reconciled against the storage provider in the same way as storage capacity resizing. The status of a PVC will reflect the currently provisioned IO as reported by the storage provider (in the same way that the status states the current size of the volume during a resize operation).
+
+```
+// Example:
+kind: PersistentVolumeClaim
+...
+spec:
+  ...
+  resources:
+requests:
+  storage: 1024Gi
+  throughput: 100 // MiB/s as unit
+  iops: 1000     // IO/s as unit
```

Comment:
for throughput and iops, does this mean "at least" this much, "exactly" this much, or "approximately" this much? Currently, the "requests" sections on Pod and PVC means "at least" this much. We could have a new "attributes" section.

Reply:
In current providers(AWS, GCP, Azure) context, the definition is that these resources are reserved and guaranteed, which means if you configure the workload correctly, this value should be exactly.

Reply:
ack on adding a new `attributes` optional field, it could be an arbitary string, e.g. there is a requirement, on aws or azure, user wants to change volume type, so if we have following new optional field, that would be great.
```
spec:
  ...
  resources:
requests:
  storage: 1024Gi
  throughput: 100 // MiB/s as unit
  iops: 1000     // IO/s as unit
  attributes: "volume_type: gp3"  // optional, change volume type to gp3
```